### PR TITLE
Feature/oauth2 jwt as access token nk ui

### PR DESCRIPTION
**Pull request by krzpiesiewicz**

### CODE REVIEW

File: ui/client/containers/NussknackerInitializer.tsx [hunk 1]
```diff
@@ -40,7 +42,9 @@ class NussknackerInitializer extends React.Component<Props, State> {
   public static ACCESS_TOKEN_CODE = 1024
 
   redirectToAuthorizeUrl(settings: AuthenticationSettings) {
-    window.location.replace(`${settings.authorizeUrl}`)
+    window.localStorage.setItem("nonce", nanoid())
+    console.log(`${settings.authorizeUrl}&nonce=${window.localStorage.getItem("nonce")}`)
```
Comment (lciolecki) [inline ui/client/containers/NussknackerInitializer.tsx:46]: Please remove console.log

File: ui/client/containers/NussknackerInitializer.tsx [hunk 2]
```diff
@@ -40,7 +42,9 @@ class NussknackerInitializer extends React.Component<Props, State> {
   public static ACCESS_TOKEN_CODE = 1024
 
   redirectToAuthorizeUrl(settings: AuthenticationSettings) {
-    window.location.replace(`${settings.authorizeUrl}`)
+    window.localStorage.setItem("nonce", nanoid())
```
Comment (lciolecki) [inline ui/client/containers/NussknackerInitializer.tsx:45]: Let's look at SystemUtils and move using localStorage to this class.

File: ui/client/containers/NussknackerInitializer.tsx [hunk 3]
```diff
@@ -136,6 +140,22 @@ class NussknackerInitializer extends React.Component<Props, State> {
 
       const queryHashParams = queryString.parse(this.props.history.location.hash)
       if (settings.implicitGrantEnabled === true && queryHashParams.access_token) {
+        if (settings.jwtAuthServerPublicKey) {
+          let verificationPassed = jwt.verify(queryHashParams.access_token, settings.jwtAuthServerPublicKey)
+          if (verificationPassed) {
+            if (queryHashParams.id_token)
+              verificationPassed = jwt.verify(queryHashParams.id_token, settings.jwtAuthServerPublicKey, {nonce: window.localStorage.getItem("nonce")})
+            else if (settings.jwtIdTokenNonceVerificationRequired === true) {
+              console.log("Warning: jwt.idTokenNonceVerificationRequired=true but id_token missing in the auth server response")
```
Comment (lciolecki) [inline ui/client/containers/NussknackerInitializer.tsx:149]: console.warn?

File: ui/client/containers/NussknackerInitializer.tsx [hunk 4]
```diff
@@ -136,6 +140,22 @@ class NussknackerInitializer extends React.Component<Props, State> {
...
+              verificationPassed = undefined
+            }
+          }
+          if (!verificationPassed) {
+            console.log("JWT verification failed")
```
Comment (lciolecki) [inline ui/client/containers/NussknackerInitializer.tsx:154]: Do we need this console.log?

Comment (krzpiesiewicz) [reply ui/client/containers/NussknackerInitializer.tsx:154]: It is not necessary nor helpful (information carried by this log is useless).
But the warning `console.warn("Warning: jwt.idTokenNonceVerificationRequired=true but id_token missing in the auth server response")` might be insufficient for a user that configures nk oauth2 settings.
There could be a helpful warning for oauth2 configuration purpose; I mean ensuring that a received access token matches JWT format (without signature verification). Access token format depends on auth server configuration and request parameters.

Comment (lciolecki) [reply ui/client/containers/NussknackerInitializer.tsx:154]: So, mayby let's do it by console.warn?:)

File: ui/client/containers/NussknackerInitializer.tsx [hunk 5]
```diff
@@ -136,6 +140,22 @@ class NussknackerInitializer extends React.Component<Props, State> {
 
       const queryHashParams = queryString.parse(this.props.history.location.hash)
       if (settings.implicitGrantEnabled === true && queryHashParams.access_token) {
+        if (settings.jwtAuthServerPublicKey) {
+          let verificationPassed = jwt.verify(queryHashParams.access_token, settings.jwtAuthServerPublicKey)
```
Comment (lciolecki) [inline ui/client/containers/NussknackerInitializer.tsx:144]: I think you should do there `acessTokenVerificationResult` and down `idTokenVerificationResult` each one should be declared as cosnt.

**Review by lciolecki — APPROVED** [review]
LGTM

**Pull request merged**